### PR TITLE
Correctly track thisContainer for this-property-assignments in JS nested containers

### LINKED ISSUE #22774: this.property assignment in JS often crashes the language service
_From @pocesar on March 21, 2018 21:20_

Version 1.22.0-insider
Commit 31c30b99345e65fffdc33066726d0962dfbecfa3
Date 2018-03-21T06:17:39.707Z
Shell 1.7.12
Renderer 58.0.3029.110
Node 7.9.0
Architecture x64

Steps to Reproduce:

1. Open an ES7 file
2. Use the following jsconfig.json:

```json
{
  "compilerOptions": {
    "checkJs": true,
    "strict": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitAny": false,
    "module": "commonjs",
    "target": "es2017",
    "alwaysStrict": true,
    "moduleResolution": "node",
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitUseStrict": true,
    "strictNullChecks": false,
    "baseUrl": "./",
    "rootDir": "./",
    "allowSyntheticDefaultImports": true
  }
}
```  
3. Variables that are defined at the top of the file are reporting as "not being read" (with a wriggly green line below), `require('./middleware')` is a wrong path, when I changed it to '../../middleware', it still shows as being './middleware'

Output > Typescript is returning this error:

```
[Error  - 18:14:55] 'quickinfo' request failed with error.
Error processing request. Cannot read property 'members' of undefined
TypeError: Cannot read property 'members' of undefined
    at bindThisPropertyAssignment (C:\nodejs\node_modules\typescript\lib\tsserver.js:20673:66)
    at bindWorker (C:\nodejs\node_modules\typescript\lib\tsserver.js:20432:29)
    at bind (C:\nodejs\node_modules\typescript\lib\tsserver.js:20336:13)
    at visitNode (C:\nodejs\node_modules\typescript\lib\tsserver.js:13322:24)
    at Object.forEachChild (C:\nodejs\node_modules\typescript\lib\tsserver.js:13547:24)
    at bindEachChild (C:\nodejs\node_modules\typescript\lib\tsserver.js:19294:16)
    at bindChildrenWorker (C:\nodejs\node_modules\typescript\lib\tsserver.js:19383:21)
    at bindChildren (C:\nodejs\node_modules\typescript\lib\tsserver.js:19269:17)
    at bind (C:\nodejs\node_modules\typescript\lib\tsserver.js:20342:21)
    at bindEach (C:\nodejs\node_modules\typescript\lib\tsserver.js:19286:21)
    at visitNodes (C:\nodejs\node_modules\typescript\lib\tsserver.js:13327:24)
    at Object.forEachChild (C:\nodejs\node_modules\typescript\lib\tsserver.js:13536:24)
    at bindEachChild (C:\nodejs\node_modules\typescript\lib\tsserver.js:19294:16)
    at bindChildrenWorker (C:\nodejs\node_modules\typescript\lib\tsserver.js:19383:21)
    at bindChildren (C:\nodejs\node_modules\typescript\lib\tsserver.js:19269:17)
    at bindContainer (C:\nodejs\node_modules\typescript\lib\tsserver.js:19250:17)
    at bind (C:\nodejs\node_modules\typescript\lib\tsserver.js:20345:21)
```

PS: When doing the `Reload Window` it takes forever for the file to start being checked (30s+)

Does this issue occur when all extensions are disabled?: Yes

_Copied from original issue: Microsoft/vscode#46292_

### PR DESCRIPTION
**Pull request by sandersn**
Previously thisContainer (previously called containerContainer) would update on every block, not just those that could bind `this`.

In addition, if the constructor symbol still can't be found, then no binding happens. This is usually OK because people don't add new properties in methods too often.

Fixes #22774

### CODE REVIEW

File: tests/baselines/reference/typeFromPropertyAssignment21.errors.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+tests/cases/conformance/salsa/chrome-devtools-DOMExtension.js(2,1): error TS2304: Cannot find name 'Event'.
+
+
+==== tests/cases/conformance/salsa/chrome-devtools-DOMExtension.js (1 errors) ====
+    // Extend that DOM!
+    Event.prototype.removeChildren = function () {
+    ~~~~~
+!!! error TS2304: Cannot find name 'Event'.
```
Comment (weswigham) [inline tests/baselines/reference/typeFromPropertyAssignment21.errors.txt]: Maybe add `@lib: dom` to this test?

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -479,11 +479,9 @@ namespace ts {
             // and block-container.  Then after we pop out of processing the children, we restore
             // these saved values.
             const saveContainer = container;
-            const saveContainerContainer = containerContainer;
+            const saveContainerContainer = thisParentContainer;
```
Comment (andy-ms) [inline src/compiler/binder.ts]: Rename this too

File: tests/cases/conformance/salsa/typeFromPropertyAssignment20.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+// @allowJs: true
+// @checkJs: true
+// @noEmit: true
+// @Filename: bluebird.js
+
+!function outer (f) { return f }(
```
Comment (andy-ms) [inline tests/cases/conformance/salsa/typeFromPropertyAssignment20.ts]: Could this be made simpler? Why are we applying identity and then negating it?

File: tests/cases/conformance/salsa/typeFromPropertyAssignment20.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+        function Async() {
+            this._trampolineEnabled = true;
+        }
+
+        Async.prototype.disableTrampolineIfNecessary = function dtin(b) {
+            if (b) {
+                this._trampolineEnabled = false;
+            }
+        };
+        var a = new Async()
```
Comment (andy-ms) [inline tests/cases/conformance/salsa/typeFromPropertyAssignment20.ts]: Are these 3 lines necessary to the test?

File: tests/cases/conformance/salsa/typeFromPropertyAssignment22.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+// @noEmit: true
+// @Filename: npm-install.js
+function Installer (where, dryrun, args, opts) {
+  this.args = args
+}
+Installer.prototype.loadArgMetadata = function (next) {
+  // ArrowFunction isn't treated as a this-container
+  (args) => {
+    this.args = args
+      this.newProperty = 1
```
Comment (andy-ms) [inline tests/cases/conformance/salsa/typeFromPropertyAssignment22.ts]: Include a test that `new Installer().newProperty` is `number | undefined`?
Also, fix indents

File: tests/cases/conformance/salsa/typeFromPropertyAssignment22.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// @allowJs: true
+// @checkJs: true
+// @noEmit: true
+// @Filename: npm-install.js
+function Installer (where, dryrun, args, opts) {
```
Comment (andy-ms) [inline tests/cases/conformance/salsa/typeFromPropertyAssignment22.ts]: Would be simpler to just write `this.args = 0` and take no parameters?

**Comment by mhegazy** [comment]
@sandersn please port to release-2.8 as well.

**Comment by sandersn** [comment]
@mhegazy @andy-ms The break introduced in #22643 is not in 2.8, so this fix does not apply to 2.8. 

To port to 2.8, I'd have to port #22643 too, and it's a complex fix to a subtle problem. (The problem is that there are too many symbols created from this-assignments, which breaks rename and find-all-refs.)

File: tests/cases/conformance/salsa/typeFromPropertyAssignment20.ts (hunk 4 above)
Comment (sandersn) [reply tests/cases/conformance/salsa/typeFromPropertyAssignment20.ts]: No, this really only checks (1) lack of crashes and (2) symbol correctness. Removed.

File: tests/cases/conformance/salsa/typeFromPropertyAssignment20.ts (hunk 3 above)
Comment (sandersn) [reply tests/cases/conformance/salsa/typeFromPropertyAssignment20.ts]: (I'm doing some digging to find out why all these pieces appear to be required.)

File: tests/cases/conformance/salsa/typeFromPropertyAssignment22.ts (hunk 5 above)
Comment (sandersn) [reply tests/cases/conformance/salsa/typeFromPropertyAssignment22.ts]: Hm. Turns out `newProperty: number | undefined` but `args: number` even when it's initialized with `this.args = 0`. I'll have to find out why.

Comment (sandersn) [reply tests/cases/conformance/salsa/typeFromPropertyAssignment22.ts]: Ah. The assignment in `loadArgMetadata` acts as a second declaration, so we union the types together. This is not ideal &mdash; we shouldn't redeclare properties that *know* are in the constructor &mdash; but I'll fix it separately.

Comment (sandersn) [reply tests/cases/conformance/salsa/typeFromPropertyAssignment22.ts]: Opened #22792 to track this; there are couple of things to consider before changing this.

File: src/compiler/binder.ts [hunk 7]
```diff
@@ -479,7 +479,7 @@ namespace ts {
             // and block-container.  Then after we pop out of processing the children, we restore
             // these saved values.
             const saveContainer = container;
-            const saveContainerContainer = thisParentContainer;
+            const saveThisContainer = thisParentContainer;
```
Comment (andy-ms) [inline src/compiler/binder.ts]: Nit: `saveThisParentContainer`, since the `thisContainer` would just be `container` usually.

**Comment by mhegazy** [comment]
thanks for the clarification. if the break is not in 2.8 we do not need to port anything.

**Pull request merged**